packages/core/src/types.ts

```diff
@@ -2368,3 +2354,40 @@ export type ToChildren<TActor extends ProvidedActor> =
               ? 'include'
               : 'exclude']
       >;
+
+export type StateConfig2 = {
+  states?: {
+    [K: string]: StateConfig2;
+  };
+};
+
+export type StateValueFromConfig<TConfig extends StateConfig2> =
+  TConfig['states'] extends Record<string, any>
+    ?
+        | keyof TConfig['states']
+        | {
+            [K in keyof TConfig['states']]?: StateValueFromConfig<
+              TConfig['states'][K]
+            >;
+          }
+    : never;
+
+export type StateValueFrom2<T extends StateConfig2> = IsAny<T> extends true
```

Comment:
it would be *very* nice to rewrite this as a tail-recursive conditional type - but we can optimize this later on